### ISSUE TITLE
Add rules engine listener

### ISSUE THREAD

**Issue by cemo** [j-easy/easy-rules#121]
I would like to fetch some data from database and pass into per rule. I noticed that there is `RuleListener` but it seems it is triggered per rule. I think that an API for per execution.

**Comment by benas** [comment]
Hi,

Yes the rule listener is executed per rule.

What do you mean an API per execution? Could you please give an example?

Kr
Mahmoud

**Comment by cemo** [comment]
I am in the process of implementing coupon campaigns. I would like to query database coupons before evaluations starting. I will pass them to rules as a fact. Specifically I need some hooks in `com.clovify.mm.core.jeasy.LifeCycleAwareRulesEngine#apply` before `LOGGER.info("Rules evaluation started");`.

**Comment by benas** [comment]
Thank you for the example, I see now. Basically you are looking for something like a `RulesEngineListener` which is executed per rule set, not per rule. I'm thinking of:

```java
public interface RulesEngineListener {
    void beforeFiringRules(Rules rules, Facts facts);
    void afterFiringRules(Rules rules, Facts facts);
}
```

With this listener, you can add custom behaviour before/after firing the whole rule set.

Is this correct? I want to make sure I correctly understand your request.

> What is the best place to fetch some data for rules?

The best place would be in the `RulesEngineListener#beforeFiringRules` method 😄

**Comment by cemo** [comment]
Yeah exactly :) By the way I had added almost same class :)

```
public interface LifeCycleListener {
  void beforeEvaluates(Rules rules, Facts facts);
  void afterEvaluates(Rules rules, Facts facts);
}```

But `RulesEngineListener` reflects better need. :)

**Comment by benas** [comment]
ok cool. But in hindsight, do we really need such a listener? What would be the added value compared to something like:

```java
Data data = .. // fetch data before rule set execution
facts.put("data", data);

rulesEngine.fire(rules, facts);

// check facts after rule set execution
```

The `RuleListener` makes sense since we don't have access to facts before/after execution of each rule (unless there is only one registered rule of course). But in the case of `RulesEngineListener`, we have access to the facts before/after the whole execution. Why would I need to implement an listener, register it and make the engine call it for me instead of writing the code before/after calling the engine as in the example. Do you see the point?

**Comment by cemo** [comment]
I will have some modules which will encapsulated all logic. Such an interface will be a contract between modules and core logic.

**Comment by benas** [comment]
That makes perfect sense. The interface actually has some benefits:

* may serve as contract between different modules (as you said)
* may encapsulate some logic that can be reused
* is more elegant than putting logic before/after running the engine (as in my previous comment)
* is consistent with the existing API: having a `RuleListener` per rule and a `RulesEngineListener` per rule set makes sense

So let me plan this for next release!
I'll let you know when a snapshot version with the addition is published to give it a try before the release.

Thank you for this suggestion!

Kr,
Mahmoud

**Comment by cemo** [comment]
@benas 

This commit is nice. The code reside in https://github.com/j-easy/easy-rules/blob/42251de30ee585c221434ec8b1f1eb8555b7538b/easy-rules-core/src/main/java/org/jeasy/rules/core/DefaultRulesEngine.java#L101-L109 can be refactored into a listener. Instead of `apply` method,  you can just use fire method. This will be nice since we can delegate in this case to default implementation.

**Comment by benas** [comment]
yes that was coming 😄 I was planning to add a `DefaultRulesEngineListener` (like the current `DefaultRuleListener`) to encapsulate the logging. This will remove the need for the apply method.

I will deploy v3.1 snapshot and keep you posted.
Thanks for the review!

**Comment by benas** [comment]
This issue should be fixed now: methods were renamed to be able to use them in both `fire` and `check` methods. A default rules engine listener was added to encapsulate the logic.

Version 3.1.0-SNAPSHOT is [deployed](https://github.com/j-easy/easy-rules/wiki/getting-started#use-a-snapshot-version) to maven central, @cemo can you give a try to the listener you requested?

**Comment by cemo** [comment]
tested. :shipit:

**Comment by benas** [comment]
Great! thank you for your time!
